sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -167,9 +175,9 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
   override lazy val dataType: MapType = {
     MapType(
       keyType = TypeCoercion.findCommonTypeDifferentOnlyInNullFlags(keys.map(_.dataType))
-        .getOrElse(StringType),
+        .getOrElse(defaultElementType),
```

Comment:
In `ArrayBasedMapBuilder`, we are explicitly asserting that `keyType` can't be `NullType`.  So assigning `NullType` here is causing test failures. 
@cloud-fan

Reply:
Can we remove that asserting?

Reply:
After removing the assertion we need to handle this as well  #23124(https://github.com/apache/spark/pull/23124/commits/9df627465d9f9bcdde129baf1304035f082e5ab6?file-filters%5B%5D=.scala#r235852779)

Reply:
yea we need to handle it. Let's add `NullType` to `case _: AtomicType | _: CalendarIntervalType if !keyType.isInstanceOf[BinaryType]`

Reply:
Yeah. I updated the PR using that approach only. All the Other review comments are handled as well.